agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -882,38 +883,53 @@ func getMetadataHealthCheck(dockerContainer *docker.Container) apicontainer.Heal
 
 // Listen to the docker event stream for container changes and pass them up
 func (dg *dockerGoClient) ContainerEvents(ctx context.Context) (<-chan DockerContainerChangeEvent, error) {
-	client, err := dg.dockerClient()
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return nil, err
 	}
-	dockerEvents := make(chan *docker.APIEvents, dockerEventBufferSize)
-	events := make(chan *docker.APIEvents)
+
+	event := make(chan *events.Message)
 	buffer := NewInfiniteBuffer()
 
-	err = client.AddEventListener(dockerEvents)
-	if err != nil {
-		seelog.Errorf("DockerGoClient: unable to add a docker event listener: %v", err)
-		return nil, err
-	}
-	go func() {
-		<-ctx.Done()
-		client.RemoveEventListener(dockerEvents)
-	}()
+	derivedCtx, cancel := context.WithCancel(ctx)
+	dockerEvents, eventErr := client.Events(derivedCtx, types.EventsOptions{})
+
+	// Cache the event from docker client. Channel closes when an error is passed to eventErr.
+	go buffer.StartListening(derivedCtx, dockerEvents)
+
+	// Receive errors from channels. If error thrown is not EOF, log and reopen channel.
+		go func(){
+			for {
+				err := <-eventErr
+				// Need to cancel context to close the stream
+				cancel()
```

Comment:
In the case of not EOF, should we cancel previous derived ctx after `StartListening` call so that we don't lose any events just in case?